libraries/chain/finality/qc.cpp

```diff
@@ -21,6 +21,80 @@ inline std::vector<uint32_t> bitset_to_vector(const vote_bitset_t& bs) {
    return r;
 }
 
+// returns true if vote indicated by active_vote_index in active_policy
+// is the same as vote indicated by pending_vote_index in pending_policy
+bool qc_t::vote_same_at(uint32_t active_vote_index, uint32_t pending_vote_index) const {
+   assert(pending_policy_sig);
+   return active_policy_sig.vote_same_at(*pending_policy_sig,
+                                         active_vote_index,
+                                         pending_vote_index);
+}
+
+// returns true if ithe other and I vote iin the same way on my_vote_index
+// and other's index.
+bool qc_sig_t::vote_same_at(const qc_sig_t& other, uint32_t my_vote_index, uint32_t other_vote_index) const {
+   assert(!strong_votes || my_vote_index < strong_votes->size());
+   assert(!weak_votes || my_vote_index < weak_votes->size());
+
+   // I vote strongly on my_vote_index, the other must vote strongly
+   // on other_vote_index too
+   if (strong_votes && (*strong_votes)[my_vote_index]) {
+      return other.strong_votes && (*other.strong_votes)[other_vote_index];
```

Comment:
Don't we also need to verify other didn't also vote weak at the same index?

Reply:
For a particular finalizer (identified by index), it can only vote strong on both active (I) and pending (other), or vote weak on both active and pending. When we know I votes strong, we only need to check if the other votes strong or not; if the other votes strong then returns true; and if the other does not vote on strong then returns false (not matter if the other votes on weak or not). Will update the comments if this makes sense.

Reply:
But if someone crafts an invalid QC they could create one that has both a strong and vote index marked as true.

Reply:
> For a particular finalizer (identified by index), it can only vote strong on both active (I) and pending (other), or vote weak on both active and pending. 

I'm not sure I follow. For a given index, the active and pending policy may have a different finalizer. There is no guarantee that any finilizer votes on both policies.

Reply:
We have already checked `qc_sig_t` that strong and weak cannot be set at the same index.

Reply:
> > For a particular finalizer (identified by index), it can only vote strong on both active (I) and pending (other), or vote weak on both active and pending.
> 
> I'm not sure I follow. For a given index, the active and pending policy may have a different finalizer. There is no guarantee that any finalizer votes on both policies.

Yes, active and pending policies can have different indexes for a same finalizer.

Reply:
Oh I see, we should continue to check the other should not vote on weak if I vote on strong -- we should eosio_assert in this case.